compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    op.getReductionDims(dims);
+    AffineMap map = op.getTiedIndexingMap(op.getInputOperand(0));
+    unsigned lastIdx = map.getNumResults() - 1;
+    unsigned lastDim = map.getDimPosition(lastIdx);
+    if (lastDim != dims[0])
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+
+    linalg::ControlSplitReductionFn fn = [this, lastIdx](linalg::LinalgOp op) {
+      return linalg::SplitReductionOptions{size, lastIdx, true};
+    };
+
+    auto opNumLoops = op.getNumLoops();
+
+    // 1) Tile to extract a single vector-length array.
+    SmallVector<int64_t> tileSizesSVFirst(opNumLoops, 1);
+    tileSizesSVFirst[opNumLoops - 1] = 0;
+    ArrayRef<int64_t> tileSizesFirst(tileSizesSVFirst);
+    linalg::LinalgTilingOptions optionsFirst;
+    optionsFirst.setTileSizes(tileSizesFirst);
+    FailureOr<linalg::TiledLinalgOp> tileResFirst =
+        linalg::tileLinalgOp(rewriter, op, optionsFirst);
+    if (failed(tileResFirst)) return failure();
+    rewriter.replaceOp(op, tileResFirst->tensorResults);
+    filter.replaceLinalgTransformationFilter(rewriter, tileResFirst->op);
+
+    // 2) Apply splitReduction on the single vector-length array. splitReduction
+    // already replaces the op.
+    FailureOr<linalg::SplitReductionResult> splitRes =
+        splitReduction(rewriter, tileResFirst->op, fn);
+    if (failed(splitRes)) return failure();
+    filter.replaceLinalgTransformationFilter(rewriter, splitRes->splitLinalgOp);
+    filter.replaceLinalgTransformationFilter(rewriter,
+                                             splitRes->resultCombiningLinalgOp);
+
+    // 3) Tile the first op generated by splitReduction with tile size of 1, to
+    // essentially create a reduction loop.
+    SmallVector<int64_t> tileSizesSV(opNumLoops + 1, 0);
+    tileSizesSV[opNumLoops - 1] = 1;
+    ArrayRef<int64_t> tileSizes(tileSizesSV);
+    linalg::LinalgTilingOptions options;
+    options.setTileSizes(tileSizes);
+    FailureOr<linalg::TiledLinalgOp> tileRes =
+        linalg::tileLinalgOp(rewriter, splitRes->splitLinalgOp, options);
+    if (failed(tileRes)) return failure();
+    rewriter.replaceOp(splitRes->splitLinalgOp, tileRes->tensorResults);
+    filter.replaceLinalgTransformationFilter(rewriter, tileRes->loops[0]);
+
+    // // 4) Unroll the tiled op (scf.for) by unrollFactor. loopUnrollByFactor
+    // // already replaces the op.
+    // LogicalResult unrolledRes = mlir::loopUnrollByFactor(
+    //     cast<scf::ForOp>(tileRes->loops[0]), unrollFactor);
+    // if (failed(unrolledRes)) return failure();
+
+    return success();
+  }
+
+ private:
+  int64_t size;
+  uint64_t unrollFactor;
+  linalg::LinalgTransformationFilter filter;
+};
+}  // namespace
+
+void LinalgSplitReductionPass::runOnOperation() {
+  func::FuncOp funcOp = getOperation();
+  int64_t useSize = size.getValue();
+  uint64_t unroll = unrollFactor.getValue();
+  if (unroll == 0) unroll = 1;
+  bool doSplitReduction = true;
+  if (useSize == 0) {
```

Comment:
Let's return directly if `useSize` is zero.

Reply:
It's not returning - it's the default, and I should run `getSplitReductionOptionsFromConfig` in that case.